### ISSUE TITLE
How to revoke refresh tokens ?

### ISSUE THREAD

**Issue by Smana** [coreos/dex#1203]
Hey guys,

I managed to setup my kubernetes cluster authentication with Dex (ldap connector) and [dex-k8s-authenticator](https://github.com/mintel/dex-k8s-authenticator) and everything works like a charm, users are able to get their credentials and i configured RBAC's on kubernetes based on ldap groups.
However i didn't found a way to revoke the refresh tokens. Is there a documentation somewhere please ?
Actually i want to enforce some policy, for instance in production the credentials have to be refreshed every 2 hours ...

Great job by the way, i love dex :)

Thank you,
Smana

**Comment by totahuanocotl** [comment]
Do you want your users to have to login every two hours? 
You could omit the refresh token from the returned response, and configure the expiry of your production cluster tokens to be two hours. 
We have a similar setup and we use [osprey](https://github.com/sky-uk/osprey) as a CLI  to login to our clusters using an LDAP backed Dex. Osprey generates the kubectl Config file for you, with no refresh token.

**Comment by Smana** [comment]
Thanks, @totahuanocotl for your answer.
Actually ideally I would like to keep refresh tokens and manage when I want to revoke them.
I'll dig further into the code.

**Comment by Smana** [comment]
I'm trying to use the grpc-client example in order to *list refresh tokens* then *delete* them according to the cluster (client_id) but i'm stuck there too.

I thought that listing the refresh tokens would be as simple as listing the passwords, but the response is empty even though there are refreshtokens (looking at k8s CRD's) :
```
func listRefreshTokens(cli api.DexClient) error {
	resp, err := cli.ListRefresh(context.TODO(), &api.ListRefreshReq{})
	if err != nil {
		return fmt.Errorf("failed to list refresh: %v", err)
	}

	log.Print("Listing Refresh:\n")
	for _, pass := range resp.RefreshTokens {
		log.Printf("%+v", pass)
	}
	return nil
}
```

Revoking a given refresh with the following piece of code
```
func revokeRefreshTokens(cli api.DexClient) error {
	ctx := context.Background()
	revokeReq := api.RevokeRefreshReq{
		UserId:   "foo",
		ClientId: "cluster1",
	}

	resp, err := cli.RevokeRefresh(ctx, &revokeReq)
	if err != nil {
		log.Fatalf("Unable to revoke refresh tokens for user: %v", err)
	}
	if resp.NotFound {
		log.Fatal("refresh token session wasn't found")
	}
	return nil
}
```
Returns an error and i don't know why yet
```
2018/03/26 06:46:04 Unable to revoke refresh tokens for user: rpc error: code = Unknown desc = illegal base64 data at input byte 8
```

digging further

**Comment by Smana** [comment]
Well maybe listing the refresh tokens is not possible because I'm using a kubernetes storage?
*storage/kubernetes/storage.go*
```
func (cli *client) ListRefreshTokens() ([]storage.RefreshToken, error) {   
        return nil, errors.New("not implemented")                                    
}
```
Any other idea to revoke the refresh tokens would be welcome :)

**Comment by Smana** [comment]
I switched to the etcd storage but it didn't help even though the functions `ListRefreshTokens` and `DeleteRefresh` seem to be implemented.
I must be missing something in the above piece of code.

**Comment by Smana** [comment]
Well, I found the missing piece of code
```
	ctx := context.Background()
	subjectString, err := internal.Marshal(&internal.IDTokenSubject{
		UserId: "skahlouch",
		ConnId: "ldap",
	})
	if err != nil {
		log.Fatalf("failed to marshal offline session ID: %v", err)
	}

	revokeReq := api.RevokeRefreshReq{
		UserId:   subjectString,
		ClientId: "dev-orscale-01-dm-paris",
	}
```
Now i need to understand why i'm still getting authenticated in the target k8s cluster even though i set the token expiry time to 2m

```./main --ca-crt=ca.crt --client-crt=client.crt --client-key=client.key
2018/03/27 08:05:57 Listing Refresh:
2018/03/27 08:05:57 refresh token session wasn't found```

**Comment by Smana** [comment]
Ok i'm now able to revoke a token for a given userId.
Now i need to find a way to list all the refresh tokens.

**Comment by Smana** [comment]
i'll try to play with etcd ```etcdctl3 get --prefix dex/refresh_token/ --print-value-only | jq [.]```

**Comment by Smana** [comment]
I wrote a script that gets a list of the users from etcd and revokes refresh tokens on a regular basis.
```
./dexctl refreshtokens revoke  --clientid dev-orscale-01-dm-paris --all --expiry 20m
```
I'll consider to open a pull request in order to list the users using dex instead of calling directly etcd.

**Comment by d4nyll** [comment]
@Smana Do you mind open-sourcing your script in a Gist?

**Comment by jseadragon** [comment]
So, I've done a bit of digging here and hope that this information helps people. I am using Dex with Gangway as the login portal, backed by an LDAP service. Maybe this won't work for your particular use case, but hopefully you find it helpful.

My general expectation is the following: 
 * A user is added to an LDAP group, and shortly after, Dex should regenerate their token with the new `group`
 * A user is removed from LDAP, and shortly after, Dex should remove their groups access.

## Use case 1 - addition / demotion of permissions via LDAP groups

TL;DR: This seems to work, out of the box. 

I've set the below config (in the helm chart):
```
expiry: 
  idTokens: "1m"
```
I have two groups in LDAP:
 - `k8s-admin` which binds to `cluster-admin` permission
 - `k8s-users` which binds to `view` permission 

I've then logged in using LDAP, generated myself a login token and installed it. As I add and remove myself from groups, my access is added and removed from the cluster once I wait for the expiry time to pass.

When I am a member of the group, you see that the command works:
```
$ kubectl describe refreshtoken  rrn63a***
...
  Groups:
    k8s-admins
```

When I remove myself from all groups, and wait 1 minute, I then see that my command fails:
```
$ kubectl describe refreshtoken  rrn63a***
Error from server (Forbidden): refreshtokens.dex.coreos.com "rrn63****" is forbidden: User "https://dex.*******/dex#MyUsername" cannot get resource "refreshtokens" in API group "dex.coreos.com" in the namespace "auth"
```

You can test this easily by running this in a shell using your auth tokens, and then adding / removing yourself from groups in LDAP. This is precisely how I tested this functionality.
```
watch --interval=0.5 'kubectl get pods'
```

## Use case2 - Revocation of user from system
When a user is removed from the system, I want to revoke their token immediately (perhaps you don't want to set the `idToken` refresh rate to 1min). I've tested this by simply just deleting the `refreshtoken` object from the cluster using another admin user.

After doing this, I simply ran the command again, and now I get a  `HTTP 400`.
```
Unable to connect to the server: failed to refresh token: oauth2: cannot fetch token: 400 Bad Request
Response: {"error":"invalid_request","error_description":"Refresh token is invalid or has already been claimed by another client."}
```

In a pinch it's trivial to iterate through all the `refreshtokens`, `describe` them, and then nuke the user you want to revoke. I dont think it has to be any more complicated than that... though not being a contributor to the project I'm not sure if there are any adverse effects from doing this. I could write a shell script which does this very easily.
